feat: use file based stream configuration

jooola opened the pull request:
Fixes  #1977 #1537
Related to #1451 #1032 #513

jooola commented:
I am having a hard time allowing to load env variables for Union types and arrays of objects (an array of Union of objects). We could really use allowing to override the output stream host names using env variables, so I am unsure if we can work around this.

It's an interesting problem though, I might need some extra time.

jooola commented:
This changes is quite big and many decisions needs to be taken. I am currently stuck at decision making and not being able to move forward. This is also defines the config schema, so I prefer not miss anything.

We have to keep in mind that:
- We can leverage the power of yaml anchors to help users to configure there streams.
- Settings that cannot be changed after Liquidsoap is started should not be editable from the UI.

Some of thinking I have done:
- I am not sure if we want to drop this 'editable' feature from the UI for now and re implement it once we released v3, because it requires a lot of changes to properly be integrated in legacy. My problem would be that we might have to deprecated fields from the config file and ask people the change again. Having a way to handle both config+UI seem handy though, but not a trivial change.
- The offline message should be editable from the UI, we should be able to edit after liquidsoap started, but it isn't implemented yet.
- Streams name/description/url/genre settings seem not changeable once liquidsoap started. So we need them in the config file.  Those are annoying to configure though, so I wonder if we should provide a default template in case of undefined values or leave it to the user to set a yaml anchor. I am unsure of there purpose too, they add details to the icecast main page, and might be used in the yellow pages listing, but I don't if this feature is really used ? I don't know if user use Icecast to get the proper metadata to there website, they might use the /api instead ?
- The legacy stream settings view needs changing, but I don't know in which direction. Having the stream status available for checking is helpful and reporting errors to the UI is a good idea. We also need to present details in the UI for users to setup there streaming software, such as the input settings. Some settings will still be editable such as replaygain/offair message/banner template. I probably need a list of output stream with there status below, and some key configuration details for those streams such as mount point/format/quality/name and description/and public URL. The rest should probably be kept private. (Maybe this is the place for starting to embed a new UI :rocket: ?)
- Some stream settings are spread across the stream settings table and the preferences table, I wonder if we should merge everything together in the pref table or move the few stream settings in the pref table to the stream settings.
- For now, we will probably require an API call from liquidsoap to the API to get some of the editable settings, so we might not need to try to move every setting to the config file.

For now, what is working:
- liquidsoap starts, generate the entrypoints properly and streams properly
- playout properly fetch the listeners stats, while I noticed it will only get the stats for a single endpoint.

Work left todo is mostly on the legacy/api side, which might require large changes to the stream settings view/model.

@paddatrapper I could use your thoughts on each of these points.

paddatrapper commented:
> I am not sure if we want to drop this 'editable' feature from the UI for now and re implement it once we released v3, because it requires a lot of changes to properly be integrated in legacy. My problem would be that we might have to deprecated fields from the config file and ask people the change again. Having a way to handle both config+UI seem handy though, but not a trivial change.

Which editable features are you wanting to keep? As far as I know, all stream settings require a liquidsoap restart to reflect. If so, then having them in the config file only makes sense and we can drop the editable feature

> The offline message should be editable from the UI, we should be able to edit after liquidsoap started, but it isn't implemented yet.

I agree, but this can be implemented at a later point I think. Removing a single line from the config isn't a major breaking change. If we can find a way for the config line to remain, that is great, but I'm not too concerned about removing it later.

> Streams name/description/url/genre settings seem not changeable once liquidsoap started. So we need them in the config file. Those are annoying to configure though, so I wonder if we should provide a default template in case of undefined values or leave it to the user to set a yaml anchor. I am unsure of there purpose too, they add details to the icecast main page, and might be used in the yellow pages listing, but I don't if this feature is really used ? I don't know if user use Icecast to get the proper metadata to there website, they might use the /api instead ?

They also populate the stream metadata viewable in things like VLC. Perhaps we provide a set of templated values and the configuration file maintains a string that the user can configure with those variables? E.g. `metadata: '{ artist } - { song }'`

> The legacy stream settings view needs changing, but I don't know in which direction. Having the stream status available for checking is helpful and reporting errors to the UI is a good idea. We also need to present details in the UI for users to setup there streaming software, such as the input settings. Some settings will still be editable such as replaygain/offair message/banner template. I probably need a list of output stream with there status below, and some key configuration details for those streams such as mount point/format/quality/name and description/and public URL. The rest should probably be kept private. (Maybe this is the place for starting to embed a new UI rocket ? Only kidding haha)

I agree - show as read-only the useful settings like connection status, public URLs, input streams, and as read/write the database-kept settings. Keep the rest private

> Some stream settings are spread across the stream settings table and the preferences table, I wonder if we should merge everything together in the pref table or move the few stream settings in the pref table to the stream settings.

What settings are they?

> For now, we will probably require an API call from liquidsoap to the API to get some of the editable settings, so we might not need to try to move every setting to the config file.

We have this already from playout to get the DB values, don't we?

> How do we validate the streams settings such as audio bitrate depending on the stream format ? The stream settings UI had some validation that helped the user. We will need good documentation for users, but validation seem required somewhere for inexperienced users. Ideally we need the service/container to fails hard if stream settings are wrong.

Probably on playout startup - have explicit errors if things are wrong and fail hard

Thanks for all the work!

jooola commented:
Depends on #2039

jooola commented:
Sorry if I didn't answer you properly. I had to let this sit some day, and had the change to tackle one small problem at the time. In the end my questions found there answers.

This isn't the cleanest PR for the legacy code as I didn't want to dive too much into the php code, which scared me when I first started working on this. The StreamSettings class now query the config file instead of the database, everything else is untouched. I've removed most of the write operation in the stream settings and around (forms/...).

What is left to do is:
- update the messages sending from legacy to playout. 
- update some left over from models schema updates. 
- probably some extra cleanup.

jooola commented on legacy/application/models/RabbitMq.php:
```diff
@@ -57,6 +57,32 @@ public static function SendMessageToPypo($event_type, $md)
         self::sendMessage($exchange, 'direct', true, $data);
     }
 
+    public static function UpdateStreamSettings($md)
+    {
+        $md['event_type'] = 'stream_settings_update';
+
+        // station_name: str
+        // input_fade_transition: float
+        // message_format: MessageFormatKind
+        // message_offline: str
+        // source_input_main_streaming: bool
+        // source_input_show_streaming: bool
+        // source_schedule_streaming: bool
+
+        $eventType = 'update_station_name';
+        $md = ['station_name' => $title];
+
+        $eventType = 'update_transition_fade';
+        $md = ['transition_fade' => $fade];
+
+        $eventType = 'update_stream_format';
+        $md = ['stream_format' => $type];
+
+        $exchange = 'airtime-pypo';
+        $data = json_encode($md, JSON_FORCE_OBJECT);
+        self::sendMessage($exchange, 'direct', true, $data);
+    }
+
```
NOTE: To remove, I'll improve this in another PR.

jooola commented on legacy/application/models/StreamSetting.php:
```diff
@@ -2,70 +2,94 @@
+            $prefix . 'description' => $output['description'] ?? '',
+            $prefix . 'genre' => $output['genre'] ?? '',
+            $prefix . 'url' => $output['url'] ?? '',
+            $prefix . 'mobile' => 'false',
```
This will always disable a feature that I am unaware off, I could use some input to know what this is really for and if we want to keep it !

paddatrapper replied:
No idea what that is...

jooola replied:
Ok, lets just leave it disabled for now.

jooola commented on playout/libretime_playout/config.py:
```diff
@@ -26,3 +31,4 @@ class Config(BaseConfig):
     general: GeneralConfig
     rabbitmq: RabbitMQConfig = RabbitMQConfig()
     playout: PlayoutConfig = PlayoutConfig()
```
The liquidsoap host and port will be moved in in own configuration schema.

jooola commented on playout/libretime_playout/liquidsoap/entrypoint.liq.j2:
```diff
@@ -0,0 +1,61 @@
+set("log.file", false)
+{%- endif %}
+
+set("server.telnet", true)
+set("server.telnet.port", 1234)
+
+{% if version >= (1, 3, 3) -%}
+set("harbor.bind_addrs", ["0.0.0.0"])
+{%- else -%}
+set("harbor.bind_addr", "0.0.0.0")
+{%- endif %}
```
NOTE: Allow custom telnet server host from configuration file.

jooola commented on playout/libretime_playout/liquidsoap/entrypoint.liq.j2:
```diff
@@ -0,0 +1,61 @@
+set("harbor.bind_addr", "0.0.0.0")
+{%- endif %}
+
+off_air_meta = "{{ settings.message_offline }}"
```
NOTE: Make this variable dynamic, and allow change from playout.

jooola commented on playout/libretime_playout/liquidsoap/entrypoint.py:
```diff
@@ -1,64 +1,82 @@
+    ARTIST_TITLE = 0
+    SHOW_ARTIST_TITLE = 1
+    RADIO_SHOW = 2
 
 
-def generate_entrypoint(log_filepath: Optional[Path]):
-    attempts = 0
-    max_attempts = 10
-    successful = False
+class StreamSettings(BaseModel):
+    station_name: str
+    input_fade_transition: float
+    message_format: MessageFormatKind
+    message_offline: str
+    source_input_main_streaming: bool
+    source_input_show_streaming: bool
+    source_schedule_streaming: bool
```
Sharing those models across api and playout is not a recommended pattern, we will have to maintain models on both sides.

paddatrapper replied:
Can they not go in `shared`?

jooola replied:
This would be great, but I am afraid of few things :
-  If we want to modify a model in one service, we have to apply the change to the other of the services using it. That is doable but commits might not represent what they supposed to changes as many other parts will be edited. Being able to modify an api and only map old => new field on model setup seem easier. This is a really important in micro services architecture, but I think it can be applied in our case, as we shared if for both the api and playout.

- Sharing logic might be a good idea for small things, but I intend to use some of those models to build advanced features that are tightly integrated in the services using it, pushing everything in the shared lib will just recreate a api-client like library were we had the playout code is spread in multiple packages.

We might be able to share those models as data structure only and inherit the classes to write some extra features, but we will have to be careful not to make it hard to use in the long run.

Also, maybe I am overthinking this and we can share it and see how it goes.

paddatrapper replied:
Maybe we add it to shared for now and if things expand or get more complicated we can refactor things out?

jooola replied:
Ok !

jooola replied:
Back to this, I think we are already hitting a too complex scenario, for example the api preferences. 

For example:
https://github.com/jooola/libretime/blob/4b17894c6e98ff76fabebdc4987ef7bb09d1f4ec/api/libretime_api/core/models/preference.py#L17-L29

will be extend really soon, when we will uncomment fields. Playout don't need some of the extra fields but with the validation system, the values will still be required as they aren't marked as optional.

So there is illusion of duplication for now, but both models will diverge really soon. I don't think we should merge this into the shared lib.

In addition preferences models will be used all over the place, maybe we should not add complexity from the start.

If we want to make sure the models are in sync for the few fields they share, we should probably create tests instead.

paddatrapper replied:
Ok, sounds reasonable to keep them separate then

jooola commented on playout/libretime_playout/timeout.py:
```diff
@@ -14,27 +12,15 @@ def __init__(self):
-        first_attempt = False
+            raise Exception("Thread did not terminate")
+
+        return thread.result
```
This whole timeout decoration might have become overkill, could probably be handled with a context manager and socket timeout.

jooola replied:
Will be refactored on a future PR

jooola commented on legacy/application/configs/conf.php:
```diff
@@ -89,6 +93,77 @@ public function getConfigTreeBuilder()
+
+            // Stream outputs
+            ->arrayNode('outputs')->arrayPrototype()->children()
+            /**/->booleanNode('enabled')->defaultFalse()->end()
```
Postponed for a future PR.

paddatrapper commented on installer/config.yml:
```diff
@@ -61,3 +61,46 @@ playout:
+    admin_password: hackme
+
+  # Output streams, max outputs is 3
+  outputs:
```
Hmm... I like the idea of rolling everything into a single outputs section. We could split it up under that or add a `type` field. I've seen it done both ways and I'm not sure which is better. I think the cleanest is probably

```yaml
outputs:
  hls:
    - server:
  icecast:
    - server:
  shoutcast:
    - server:
```

Yeah, I don't know who actually uses shoutcast...

jooola replied:
Do we agree on the following config ?
```yml
  # Default output
  .default_output: &default_output
    kind: icecast
    source_user: source
    source_password: hackme
    admin_user: admin
    admin_password: hackme

    name: LibreTime!
    description: LibreTime Radio!
    website: https://libretime.org
    genre: various

  outputs:
    - <<: *default_output
      enabled: true
      mount: main.ogg
      audio:
        format: ogg
        bitrate: 256

    - <<: *default_output
      enabled: true
      mount: main.mp3
      audio:
        format: mp3
        bitrate: 256

    - kind: shoutcast
      enabled: true
      source_password: password
      audio:
        format: mp3
        bitrate: 256

    - kind: alsa
      enabled: true
```

The outputs field should never be used in python but properties filtering this field can. This is specially required when listing outputs and guessing legacy ids based on there index, as we have to build up the `s1`, `sX` values from these lists.

For PHP, we will have to preprocess the outputs list, as the symfony validation component does not handle [Union type with a discriminator field](https://github.com/jooola/libretime/blob/4b17894c6e98ff76fabebdc4987ef7bb09d1f4ec/shared/libretime_shared/config/_models.py#L191-L194) to sort them out.

jooola replied:
Hmm, going back and forth.

Why don't we want to have `stream.line_outputs`, `stream.icecast_outputs` or `stream.hls_outputs` ?

paddatrapper replied:
That was what I was suggesting, except that `_outputs` is unnecessary in the name because they are under the `outputs` key

jooola replied:
I might be bike shedding here, but I was hoping not to add too much nesting.

jooola commented on api/libretime_api/permissions.py:
```diff
@@ -98,10 +115,12 @@ def has_permission(self, request, view):
             return request.user.has_perm(perm, obj)
-        return check_authorization_header(request)
+
+        return super().has_object_permission(request, view, obj)
```
I am a bit lost with our custom permission management system. IS there a way to simplify it ?

I also wonder if we currently expose way too much data to the public, shouldn't we only allow the api key auth and admin auth for now and slowly grant permissions when needed ?

paddatrapper replied:
Currently this mimics the legacy permissions - ensuring that users have access to things via the API that they would have access to via the UI. Nothing sensitive should be publicly accessible

jooola commented:
This is ready for review and lots of testing !

jooola commented:
While the PR is "working", here is what is left todo:
- review each change and do some extra cleaning/refactoring.
- test this PR with weird configs to find uncovered edge cases.
- write more documentation once we are happy with the code.
- write an upgrade procedure.

jooola commented:
/website-preview

jooola commented on installer/config.yml:
```diff
@@ -85,3 +85,170 @@ playout:
+  inputs:
+    # Main harbor input.
+    main:
+      # Harbor input public url. If not defined, the value will be generated from
+      # the [general.public_url] hostname and the input port and mount.
+      url:
```
Unsure about the naming of this fields, maybe url is too generic. Maybe we should use public_url or stream_url ?

paddatrapper replied:
`public_url` I think is a good name here

jooola commented on installer/config.yml:
```diff
@@ -85,3 +85,170 @@ playout:
+      - # Whether the output is enabled.
+        # > default is false
+        enabled: false
+        # Output public url, If not defined, the value will be generated from
+        # the [general.public_url] hostname and the output port and mount.
+        url:
```
Same as for the input url, maybe we should have public_url ?

jooola commented on installer/config.yml:
```diff
@@ -85,3 +85,170 @@ playout:
+        # Shoutcast stream genre.
+        genre: various
+
+    # Line outputs.
+    # > max items is 1
+    line:
```
Is line outputs a good name ? Should we consider hardware output ? Unsure which name is the best to group all the system outputs.

jooola replied:
Maybe `device` or `system`. We have to group the following alsa, ao, oss, portaudio, pulseaudio.

This list could probably be extended at some point, with gstreamer, and other outputs supported by liquidsoap.

paddatrapper replied:
I like `system` , think it describes what it is better than device

jooola commented:
Unsure if we can provide much help for the upgrade procedure, hope this is enough.

Anyway, I tested this on bullseye, its ready to be reviewed.

jooola commented:
/website-preview

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
 
+## Stream
+
+The `stream` section configure anything related to the inputs and outputs streams of the project.
```
configures

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+
+:::info
+
+To help you simplify your stream configuration, you can use yaml anchors to define a common properties and reuse them in your outputs definition:
```
output definitions

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+
+### Inputs
+
+The `stream.inputs` section configure anything related to the inputs streams of the project.
```
configures

input streams

of LibreTime maybe?

paddatrapper replied:
configures

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+    # Main harbor input.
+    main:
+      # Harbor input public url. If not defined, the value will be generated from
+      # the [general.public_url] hostname and the input port and mount.
```
hostname, the input port and mount

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+    # Show harbor input.
+    show:
+      # Harbor input public url. If not defined, the value will be generated from
+      # the [general.public_url] hostname and the input port and mount.
```
hostname. the input port and mount

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+
+### Outputs
+
+The `stream.outputs` section configure anything related to the outputs streams of the project.
```
output streams of LibreTime

paddatrapper replied:
configures

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+        # > default is false
+        enabled: false
+        # Output public url, If not defined, the value will be generated from
+        # the [general.public_url] hostname and the output port and mount.
```
hostname, the output port and mount

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+        # Icecast admin user.
+        # > default is admin
+        admin_user: "admin"
+        # Icecast admin password. If not defined, stats will not be collected.
```
statistics?

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+
+#### System
+
+The `stream.outputs.system` section configure the system outputs.
```
configures

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+
+#### Shoutcast
+
+The `stream.outputs.shoutcast` section configure the shoutcast output streams.
```
configures

paddatrapper commented on docs/admin-manual/setup/configuration.md:
```diff
@@ -221,6 +221,240 @@ In order to apply the changes made in this section, please restart the following
+
+#### Icecast
+
+The `stream.outputs.icecast` section configure the icecast output streams.
```
configures

paddatrapper commented on docs/releases/unreleased.md:
```diff
@@ -36,6 +36,24 @@ Along with the Debian Buster deprecation, the following dependencies versions ar
+
+:::caution
+
+Please run this **before the upgrade procedure**!
```
Maybe `follow this`, as there aren't actually commands to run?

jooola commented:
/website-preview

paddatrapper commented:
I've done some vagrant testing on this too and looks like it is working consistently

paddatrapper commented:
Do we want to leave the settings in the DB for a release? I.e. do not use them, but do not drop the table either. On the other hand people should be making a backup and reading release notes before upgrading...

jooola commented:
> I've done some vagrant testing on this too and looks like it is working consistently

Which distro did you test ? I also started testing again this morning. With the recent bionic compatibility problems I decided to properly test on it.

> Do we want to leave the settings in the DB for a release? I.e. do not use them, but do not drop the table either. On the other hand people should be making a backup and reading release notes before upgrading...

I'm fine either way, we can postpone the table cleanup for 3.0.0. We could also dump the table in a file "in case of emergency" before the upgrade procedure.

It doesn't require much work and might help those who don't read the documentation...

paddatrapper commented:
I tested on focal and bullseye.

I think either of those options are good

jooola commented:
- Did a test on bionic, nothing obvious is broken !
- Fixed legacy to only use the config `inputs.*.public_url` if they are truthy (not empty strings)
- Remove the database migration that flushes the stream settings, and added commands that helps user to view this data.

https://github.com/libretime/libretime/compare/dd97598d74efff42bf48940cae5762015a69dd61..1f65fc9baadb245ad3afd6b3f55a8557c25bc578

paddatrapper reviewed (approved):
Sounds good!

jooola commented:
I guess we are good for merging ?
